src/transform/src/fusion/join.rs

```diff
@@ -50,85 +50,152 @@ impl Join {
         if let MirRelationExpr::Join {
             inputs,
             equivalences,
-            demand,
-            implementation,
+            ..
         } = relation
         {
-            let mut new_inputs = Vec::new();
-            let mut new_equivalences = Vec::new();
-            let mut new_columns = 0;
+            let mut join_builder = JoinBuilder::new(equivalences);
 
             // We scan through each input, digesting any joins that we find and updating their equivalence classes.
             // We retain any existing equivalence classes, as they are already with respect to the cross product.
             for input in inputs.drain(..) {
-                if let MirRelationExpr::Join {
-                    mut inputs,
-                    mut equivalences,
-                    ..
-                } = input
-                {
-                    // Update and push all of the variables.
-                    for mut equivalence in equivalences.drain(..) {
-                        for expr in equivalence.iter_mut() {
-                            expr.visit_mut(&mut |e| {
-                                if let MirScalarExpr::Column(c) = e {
-                                    *c += new_columns;
-                                }
-                            });
+                match input {
+                    MirRelationExpr::Join {
+                        inputs,
+                        equivalences,
+                        ..
+                    } => {
+                        // Merge the inputs into the new join being built.
+                        join_builder.add_inputs(inputs, equivalences, None);
+                    }
+                    MirRelationExpr::Filter { input, predicates } => {
+                        if let MirRelationExpr::Join {
+                            inputs,
+                            equivalences,
+                            ..
+                        } = *input
+                        {
+                            // Merge the inputs and the predicates into the new join being built.
+                            join_builder.add_inputs(inputs, equivalences, Some(predicates));
+                        } else {
+                            // Retain the input.
+                            let input = input.filter(predicates);
```

Comment:
I prefer not having to reconstruct the original expression.

You could change line 70 to:
```
MirRelationExpr::Filter {input: inner_input, predicates} = {
```
and eliminate line 81.

Reply:
In order to be able to do that I'd need to do the pattern matching by reference which won't allow me to pass the ownership of `equivalences` and `predicates`  to `add_inputs/add_subjoin` in the branch above, which will make me clone them. I think reconstructing transferring the ownership is better than cloning and object that will be potentially thrown away. Is there a rusty of way of avoiding both things at the same time?

Reply:
Hmm yeah the tricky thing is that the filter input field is a Box<MirScalarExpr>, so you can't do the normal rust-y way of matching just the Filter around a Join case, a.k.a.
`MirRelationExpr::Filter{ input: MirRelationExpr::Join{..}, predicates} => add_subjoin(..)`
Also matching in this manner is apparently still experimental:
`MirRelationExpr::Filter {input: inner_input, predicates} if let MirRelationExpr::Join { .. } = *inner_input=> add_subjoin(..)`.

That said, the other predicate lifting transform `JoinImplementation` does not assume that filters are fused and instead recurses into Filters. 
https://github.com/MaterializeInc/materialize/blob/79317904d725532d6736bbf66d3e830413f103fd/src/transform/src/join_implementation.rs#L110

What do you think of changing the section here to something like the below?
```
fn process_input(MirRelationExpr input) {
  match input {
    MirRelationExpr::Join { input, equivalences, ..} => join_builder.add_subjoin(input, equivalences),
    MirRelationExpr::Filter{input, predicates} => {
      join_builder.add_predicates(predicates)
      process_input(input)
    }
    _ => join_builder.add_input(input)
  }
}
```
Added predicates would be held in a Vec to be applied to the next input/subjoin added.

On the other hand, it can be argued that we should just assume the filters are fused by the time join fusion or join implementation runs, and the assumption should be added to the comment in this file.